_posts/2019-03-23-WassersteinGAN.markdown

```diff
@@ -0,0 +1,528 @@
+        * Hint, it has to do with the approximation for the JSD & the dimensionality of the data manifolds.
+        <details><summary>Solution</summary>
+        <p>
+        If we train the discriminator too much we get vanishing gradients. This is due to the fact that when the true data distribution and model distribution lie on low dimensional manifolds (or have disjoint support almost everywhere), the optimal discriminator will be perfect &#8212; i.e. the gradient will be zero almost everywhere. This is something that almost always happens.
+        </p>
+        <p>
+        On the other hand, if we train the discriminator too little, then the loss for the generator no longer approximates the JSD. This is because the approximation only holds if the discriminator is near the optimal \(D^*(x) = \frac{p)d(x)}{p_d(x) + p_g(x)}\). 
+        </p>
+        </details>
+
+  4. Implement a GAN and train it on Fashion MNIST.
+      * [This notebook](https://colab.research.google.com/drive/1OWZEeF-SB0r1f6mHm-7-hfxd2zsecEwq#scrollTo=Q8YoJ4mejp97) contains a skeleton with boilerplate code and hints.
+      * Try various settings of hyper-parameters, other than those suggested, and see if the model converges.
+      * Examine samples from various stages of the training. Rank them without looking at the corresponding loss and see if your ranking agrees with the loss.
+      <details><summary>Solution</summary>
+        <p>
+        <a href="https://github.com/eriklindernoren/Keras-GAN/blob/master/dcgan/dcgan.py">Here</a> is a GAN implementation using Keras.
+        </p>
+      </details>
+  
+  **Notes**: Here is a [link](todo:add_link) to our notes for the lesson. We were fortunate enough to have Martin Arjovsky sit in on the session!
+
+  <br />
+
+# 4 Wasserstein GAN
+
+  **Motivation**: Last week we saw how GANs solve some problems in training generative models but also that they bring in new problems. This week we’ll look at the Wasserstein GAN which goes a long way to solving these problems.
+
+  **Topics**:
+
+  1. Wasserstein Distance vs KLD/JSD
+  2. Wasserstein GAN
+
+  **Required Reading**:
+
+  1. [The WGAN paper](https://arxiv.org/pdf/1701.07875.pdf)
+      * This should be pretty self-explanatory! We’re doing a DFL on Wasserstein GANs so we’d better read the paper! (This isn’t the end of the road, however, next week we’ll look at WGAN-GP.) The paper builds upon an intuitive idea: the family of Wasserstein distances is a nice distance between probability distributions, that is well grounded in theory. The authors propose to use the 1-Wasserstein distance to estimate generative models. They show that the 1-Wasserstein distance is an IPM with a meaningful set of constraints (1-Lipschitz functions), and can, therefore, be optimized by focusing on discriminators that are “well behaved” (meaning that their output does not change to much if you perturb the input, i.e. they are Lipschitz!).
```

Comment:
I think the statement: "use the 1-Wasserstein distance to estimate generative models" should be expanded on a little bit.

Reply:
I've expanded a bit on this in my most recent commits, let me know if you think it needs more.